rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -230,97 +90,77 @@ class Publisher : public PublisherBase
+    // unique_ptr<MessageT> or shared_ptr<const MessageT> siganture could
+    // be used here. I decided to use the first one.
     auto ptr = MessageAllocTraits::allocate(*message_allocator_.get(), 1);
-    MessageAllocTraits::construct(*message_allocator_.get(), ptr, *msg.get());
-    MessageUniquePtr unique_msg(ptr, message_deleter_);
-    return this->publish(unique_msg);
+    MessageAllocTraits::construct(*message_allocator_.get(), ptr, *msg);
+    MessageUniquePtr unique_msg = MessageUniquePtr(ptr, message_deleter_);
+    this->publish(unique_msg);
   }
 
   virtual void
-  publish(std::shared_ptr<const MessageT> msg)
+  publish(const std::shared_ptr<MessageT> && msg)
   {
-    // Avoid allocating when not using intra process.
-    if (!store_intra_process_message_) {
-      // In this case we're not using intra process.
-      return this->do_inter_process_publish(msg.get());
+    // Cast to std::shared_ptr<const MessageT> and publish.
+    this->publish(std::const_pointer_cast<const MessageT>(msg));
+  }
+
+  virtual void
+  publish(const std::shared_ptr<const MessageT> & msg)
+  {
+    if (intra_process_is_enabled_) {
+      uint64_t message_seq =
+        store_intra_process_message(intra_process_publisher_id_, msg);
+      this->do_intra_process_publish(message_seq);
+    }
+    if (get_subscription_count() > get_intra_process_subscription_count()) {
+      this->do_inter_process_publish(msg.get());
     }
-    // Otherwise we have to allocate memory in a unique_ptr and pass it along.
-    // TODO(wjwwood):
-    //   The intra process manager should probably also be able to store
-    //   shared_ptr's and do the "smart" thing based on other intra process
-    //   subscriptions. For now call the other publish().
-    auto ptr = MessageAllocTraits::allocate(*message_allocator_.get(), 1);
-    MessageAllocTraits::construct(*message_allocator_.get(), ptr, *msg.get());
-    MessageUniquePtr unique_msg(ptr, message_deleter_);
-    return this->publish(unique_msg);
   }
 
   virtual void
   publish(const MessageT & msg)
   {
     // Avoid allocating when not using intra process.
-    if (!store_intra_process_message_) {
+    if (!intra_process_is_enabled_) {
       // In this case we're not using intra process.
       return this->do_inter_process_publish(&msg);
     }
     // Otherwise we have to allocate memory in a unique_ptr and pass it along.
+    // NOTE(ivanpauno): As the message is not const, a copy should be made.
+    // unique_ptr<MessageT> or shared_ptr<const MessageT> siganture could
+    // be used here. I decided to use the first one.
     auto ptr = MessageAllocTraits::allocate(*message_allocator_.get(), 1);
     MessageAllocTraits::construct(*message_allocator_.get(), ptr, msg);
-    MessageUniquePtr unique_msg(ptr, message_deleter_);
-    return this->publish(unique_msg);
+    MessageUniquePtr unique_msg = MessageUniquePtr(ptr, message_deleter_);
```

Comment:
What's the difference here from the original `MessageUniquePtr unique_msg(ptr, message_deleter_);`?

Reply:
I probably deleted and re-wrote it differently. I will come back to the original way, as it's cleaner.